pkgs/development/libraries/glibc/ld-fallback.patch

```diff
@@ -0,0 +1,275 @@
+
+diff --git a/elf/dl-load.c b/elf/dl-load.c
+index 9a87fda9c9..a8fd639238 100644
+--- a/elf/dl-load.c
++++ b/elf/dl-load.c
+@@ -101,6 +101,8 @@ int __stack_prot attribute_hidden attribute_relro
+ 
+ /* This is the decomposed LD_LIBRARY_PATH search path.  */
+ struct r_search_path_struct __rtld_env_path_list attribute_relro;
++/* This is the decomposed LD_FALLBACK_PATH search path.  */
++struct r_search_path_struct __rtld_fallback_path_list attribute_relro;
+ 
+ /* List of the hardware capabilities we might end up using.  */
+ #ifdef SHARED
+@@ -701,6 +703,7 @@ cache_rpath (struct link_map *l,
+ 
+ void
+ _dl_init_paths (const char *llp, const char *source,
++		const char *lfp, const char *fsource,
+ 		const char *glibc_hwcaps_prepend,
+ 		const char *glibc_hwcaps_mask)
+ {
+@@ -854,6 +857,45 @@ _dl_init_paths (const char *llp, const char *source,
+     }
+   else
+     __rtld_env_path_list.dirs = (void *) -1;
++
++  /* Made by copying the above followed by
++      %s/llp/lfp/g
++      %s/env_path_list/fallback_path_list/g
++      %s/source/fsource/g
++   */
++  if (lfp != NULL && *lfp != '\0')
++    {
++      char *lfp_tmp = strdupa (lfp);
++
++      /* Decompose the LD_LIBRARY_PATH contents.  First determine how many
++	 elements it has.  */
++      size_t nlfp = 1;
++      for (const char *cp = lfp_tmp; *cp != '\0'; ++cp)
++	if (*cp == ':' || *cp == ';')
++	  ++nlfp;
++
++      __rtld_fallback_path_list.dirs = (struct r_search_path_elem **)
++	malloc ((nlfp + 1) * sizeof (struct r_search_path_elem *));
++      if (__rtld_fallback_path_list.dirs == NULL)
++	{
++	  errstring = N_("cannot create cache for search path");
++	  goto signal_error;
++	}
++
++      (void) fillin_rpath (lfp_tmp, __rtld_fallback_path_list.dirs, ":;",
++			   fsource, NULL, l);
++
++      if (__rtld_fallback_path_list.dirs[0] == NULL)
++	{
++	  free (__rtld_fallback_path_list.dirs);
++	  __rtld_fallback_path_list.dirs = (void *) -1;
++	}
++
++      __rtld_fallback_path_list.malloced = 0;
++    }
++  else
++    __rtld_fallback_path_list.dirs = (void *) -1;
++
+ }
+ 
+ 
+@@ -1928,9 +1970,10 @@ open_path (const char *name, size_t namelen, int mode,
+       if (sps->malloced)
+ 	free (sps->dirs);
+ 
+-      /* __rtld_search_dirs and __rtld_env_path_list are
++      /* __rtld_search_dirs, __rtld_env_path_list,
++	 and __rtld_fallback_path_list are
+ 	 attribute_relro, therefore avoid writing to them.  */
+-      if (sps != &__rtld_search_dirs && sps != &__rtld_env_path_list)
++      if (sps != &__rtld_search_dirs && sps != &__rtld_env_path_list && sps != &__rtld_fallback_path_list)
+ 	sps->dirs = (void *) -1;
+     }
+ 
+@@ -2160,6 +2203,17 @@ _dl_map_object (struct link_map *loader, const char *name,
+ 	}
+ #endif
+ 
++      /* Try the LD_FALLBACK_PATH environment variable.  */
++      /* Made by copying LD_LIBRARY_PATH snippet above followed by
++          %s/env_path_list/fallback_path_list/g
++          %s/LA_SER_LIBPATH/LA_SER_DEFAULT/g
```

Comment:
Why this last substitution?

This tag is used for audits...  `LA_SER_LIBPATH` means "I got this path from an environment variable", which is probably the most accurate description for libraries loaded via `LD_FALLBACK_PATH`.

Reply:
Hi! I'm pondering at the thought that maybe this PR (and generally, some kind of glibc patching) could improve the situation around `import torch` in a FHS (Ubuntu) environment, where PyTorch is built from nixpkgs with ROCm/CUDA support. At the very least, it could make tools like `numtide/nix-gl-host` "safer" by making it more explicit that the override should only affect a selection of libraries, but none of the libraries linked by Nix through the Runpaths.

And I'm re-iterating, but `makeWrapper` does not (reasonably) address `import torch`